### PR TITLE
HybridFactorGraph base class and other niceties

### PR DESCRIPTION
**Pull request by varunagrawal**
Made `HybridFactorGraph` a base class for `HybridGaussianFactorGraph` and for `HybridNonlinearFactorGraph` which will be added in a later PR. This way, the code is more DRY and mirrors the inheritance structure of the `[Nonlinear|Gaussian]FactorGraph` classes.

Refactored tests for `HybridGaussianFactorGraph` to make them much more descriptive and to actually test things.

Also added tests and improvements to `HybridBayesNet`.

### CODE REVIEW

File: gtsam/hybrid/HybridGaussianFactorGraph.h [hunk 1]
```diff
@@ -72,10 +73,16 @@ struct EliminationTraits<HybridGaussianFactorGraph> {
...
+    : public HybridFactorGraph,
       public EliminateableFactorGraph<HybridGaussianFactorGraph> {
+ protected:
+  /// Check if FACTOR type is derived from GaussianFactor.
+  template <typename FACTOR>
+  using IsGaussian = typename std::enable_if<
+      std::is_base_of<GaussianFactor, FACTOR>::value>::type;
+
  public:
-  using Base = FactorGraph<HybridFactor>;
```
Comment (ProfFan) [inline gtsam/hybrid/HybridGaussianFactorGraph.h:78]: nice!

File: gtsam/hybrid/HybridGaussianFactorGraph.h [hunk 2]
```diff
@@ -113,6 +126,40 @@ class GTSAM_EXPORT HybridGaussianFactorGraph
...
   void add(boost::shared_ptr<DecisionTreeFactor> factor);
+
+  /**
+   * Add a gaussian factor *pointer* to the internal gaussian factor graph
+   * @param gaussianFactor - boost::shared_ptr to the factor to add
+   */
+  template <typename FACTOR>
+  IsGaussian<FACTOR> push_gaussian(
+      const boost::shared_ptr<FACTOR>& gaussianFactor) {
+    Base::Base::push_back(
```
Comment (ProfFan) [inline gtsam/hybrid/HybridGaussianFactorGraph.h:137]: This looks a bit strange

File: gtsam/hybrid/tests/testGaussianHybridFactorGraph.cpp [hunk 3]
```diff
@@ -273,26 +266,36 @@ TEST(HybridGaussianFactorGraph, eliminateFullMultifrontalTwoClique) {
...
-  on X(1), which is not possible because it involves eliminating discrete before
-  continuous. The solution to this, however, is in Murphy02. TLDR is that this
-  is 1. expensive and 2. inexact. neverless it is doable. And I believe that we
-  should do this.
+  (Fan) Explanation: the Junction tree will need to reeliminate to get to the
+  marginal on X(1), which is not possible because it involves eliminating
+  discrete before continuous. The solution to this, however, is in Murphy02.
+  TLDR is that this is 1. expensive and 2. inexact. nevertheless it is doable.
+  And I believe that we should do this.
   */
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testGaussianHybridFactorGraph.cpp:282]: I think originally there is a `// hbt->marginalFactor(X(1))->print("HBT: ");`, which is what this explanation is for

**Review by ProfFan — COMMENTED** [review]
some nits, in general looks good to me.

will wait on Frank for the API

File: gtsam/hybrid/HybridGaussianFactorGraph.h (hunk 2 above)
Comment (ProfFan) [reply gtsam/hybrid/HybridGaussianFactorGraph.h:137]: I think just use the base class name is clearer here

File: gtsam/hybrid/HybridGaussianFactorGraph.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.h:78]: Yeah, this makes adding `HybridNonlinearFactorGraph` later a cinch! :D

File: gtsam/hybrid/tests/testGaussianHybridFactorGraph.cpp (hunk 3 above)
Comment (varunagrawal) [reply gtsam/hybrid/tests/testGaussianHybridFactorGraph.cpp:282]: Okay I guess that was the `hbt->dot` statement. Thanks for letting me know!

File: gtsam/hybrid/HybridGaussianFactorGraph.h (hunk 2 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.h:137]: Great catch!

**Comment by varunagrawal** [comment]
@ProfFan Frank is travelling and won't be available to review. I can leave this open until he gets back since I am pivoting to something else. :)

**Comment by dellaert** [comment]
Can you add more info **in the PR comment** as to the base class that is introduced, and which classes derive from it, and *why* this change was made?

**Comment by varunagrawal** [comment]
@dellaert done

**Comment by varunagrawal** [comment]
@dellaert ping

**Comment by varunagrawal** [comment]
Some love here please?

**Comment by ProfFan** [comment]
CI is failing

**Comment by varunagrawal** [comment]
That's because of the Ubuntu 18.04 brownout. Should work now.

File: gtsam/hybrid/tests/testHybridValues.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  values2.insert(98, Vector2(2, 3));
+  EXPECT(!assert_equal(values2, values));
+}
+
+/* ************************************************************************* */
+int main() {
+  TestResult tr;
+  return TestRegistry::runAllTests(tr);
+}
+/* ************************************************************************* */
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridValues.cpp:58]: newline

File: gtsam/hybrid/HybridBayesNet.h [hunk 5]
```diff
@@ -36,6 +38,35 @@ class GTSAM_EXPORT HybridBayesNet : public BayesNet<HybridConditional> {
 
   /** Construct empty bayes net */
   HybridBayesNet() = default;
+
+  /// Add HybridConditional to Bayes Net
+  using Base::add;
+
+  /// Add a discrete conditional to the Bayes Net.
+  void add(const DiscreteKey &key, const std::string &table) {
```
Comment (ProfFan) [inline gtsam/hybrid/HybridBayesNet.h:46]: I am not in favor of more helpers, do we have similar stuff in `DiscreteBayesNet`?

File: gtsam/hybrid/tests/testHybridLookupDAG.cpp [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+  EXPECT(assert_equal(hv.at(X(2)), d2));
+  EXPECT(assert_equal(hv.at(X(1)), d2 + d1));
+}
+
+/* ************************************************************************* */
+int main() {
+  TestResult tr;
+  return TestRegistry::runAllTests(tr);
+}
+/* ************************************************************************* */
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridLookupDAG.cpp:272]: newline

**Review by ProfFan — APPROVED** [review]
Approving sans two comments

File: gtsam/hybrid/tests/testHybridLookupDAG.cpp (hunk 6 above)
Comment (varunagrawal) [reply gtsam/hybrid/tests/testHybridLookupDAG.cpp:272]: This is from @xsj01's PR. @xsj01 seems like your formatter isn't set up properly yet.

**Pull request merged**